client_plugin/drivers/vfile/vfile_driver.go

```diff
@@ -371,61 +386,69 @@ func (d *VolumeDriver) Remove(r volume.Request) volume.Response {
 		return volume.Response{Err: msg}
 	}
 
-	// Test and set status to Deleting
-	if !d.kvStore.CompareAndPutStateOrBusywait(kvstore.VolPrefixState+r.Name,
-		string(kvstore.VolStateReady),
-		string(kvstore.VolStateDeleting)) {
-		// Failed to change state from Ready to Deleting
-		// 1. Volume is in Mounted state -> get clients and return error
-		// 2. Volume is already in Deleting/Error/Creating/Unmounting and timeout -> continue delete
-		// 3. Volume is in other states -> return error
+	// Get the lock for changing the global refcount
+	grefLock, err := d.kvStore.CreateLock(kvstore.VolPrefixGRef + r.Name)
+	if err != nil {
+		msg = fmt.Sprintf("Failed to create lock for removing volume %s. Error: %v",
+			kvstore.VolPrefixGRef+r.Name, err)
+		log.Error(msg)
+		return volume.Response{Err: msg}
+	}
 
-		// Get a list of host VMs using this volume, if any
-		keys := []string{
-			kvstore.VolPrefixState + r.Name,
-			kvstore.VolPrefixInfo + r.Name,
-		}
-		entries, err := d.kvStore.ReadMetaData(keys)
-		if err != nil {
-			msg = fmt.Sprintf("Remove failed: cannot read metadata of volume %s", r.Name)
-			log.Errorf(msg)
-			return volume.Response{Err: msg}
-		}
+	err = grefLock.BlockingLockWithLease()
+	if err != nil {
+		msg = fmt.Sprintf("Failed to blocking wait lock for removing volume %s. Error: %v",
+			kvstore.VolPrefixGRef+r.Name, err)
+		log.Error(msg)
+		grefLock.ClearLock()
+		return volume.Response{Err: msg}
+	}
 
-		state := entries[0].Value
-		switch state {
-		case string(kvstore.VolStateDeleting):
-			log.Warningf("Remove: volume in Deleting state after timeout. Continue deleting.")
-		case string(kvstore.VolStateError):
-		case string(kvstore.VolStateCreating):
-		case string(kvstore.VolStateUnmounting):
-			log.Warningf("Remove: volume in %s state after timeout. Continue deleting", state)
-			if !d.kvStore.CompareAndPut(kvstore.VolPrefixState+r.Name,
-				state, string(kvstore.VolStateDeleting)) {
-				msg = fmt.Sprintf("Remove: Volume state changed unexpected. Please retry later")
-				log.Errorf(msg)
-				return volume.Response{Err: msg}
+	// Try to lock for changing state. At this point, there should be no lock on the state
+	stateLock, err := d.kvStore.CreateLock(kvstore.VolPrefixState + r.Name)
+	if err != nil {
+		msg = fmt.Sprintf("Failed to create lock for removing volume %s. Error: %v",
+			kvstore.VolPrefixState+r.Name, err)
+		log.Error(msg)
+		grefLock.ReleaseLock()
+		return volume.Response{Err: msg}
```

Comment:
I think we should call stateLock.ClearLock() here.

Reply:
No this error is when it's failed to create state lock. Since it's not created, we don't need to clear it.